syscalls_linux.go

```diff
@@ -113,6 +113,17 @@ func setDeviceOptions(fd uintptr, config Config) (err error) {
 	if config.Persist {
 		value = 1
 	}
+
 	return ioctl(fd, syscall.TUNSETPERSIST, uintptr(value))
 
 }
+
+func openTun() (*os.File, error) {
+	tunFile := "/dev/net/tun"
+	bfile, err := os.OpenFile(tunFile, os.O_RDWR, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	return os.NewFile(bfile.Fd(), tunFile), nil
```

Comment:
I haven't got to actually try this out myself, but does this set it to non-blocking mode? It seems `os.NewFile` only SetNonBlock if the fd is already in non-blocking mode.

Or is it actually about registering with netpoll system [here](https://golang.org/src/os/file_unix.go?s=2621:2664#L139)? Sorry I'd just love to understand this better.

Reply:
@songgao what I actually want is registering with netpoll system. But sounds like the code does not do that.

Instead, when we call `bfile.Fd()`, the [Fd()](https://github.com/golang/go/blob/835f983da6866deec1c6550555c96de729c8848c/src/os/file_unix.go#L69) will put the file descriptor to blocking mode. So `os.NewFile` see it's not a pollable and does not register it with netpoll.

The `os.NewFile(bfile.Fd(), tunFile)` actually returns a file in blocking mode, so why it works but the original does not?

Reply:
@songgao ping